### PR TITLE
SONARJAVA-1425 - Member select redundant before method invocation

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java [hunk 1]
```diff
@@ -394,4 +399,22 @@ public String toString() {
       return leftOp + " ^ " + rightOp;
     }
   }
+
+  static class BooleanObjectSymbolicValue extends SymbolicValue {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/SymbolicValue.java]: MAJOR Override this superclass' "equals" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2160)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-squid/src/test/java/org/sonar/java/cfg/CFGTest.java [hunk 2]
```diff
@@ -716,15 +701,13 @@ public void foreach_loop_continue() {
...
-            element(Tree.Kind.MEMBER_SELECT),
             element(Tree.Kind.METHOD_INVOCATION)).successors(2),
         block(
             element(Tree.Kind.VARIABLE, "foo")).terminator(Tree.Kind.FOR_EACH_STATEMENT).successors(1, 5),
         block(
             element(Tree.Kind.CHAR_LITERAL, "'end'"),
             element(Tree.Kind.IDENTIFIER, "System"),
-            element(Tree.Kind.MEMBER_SELECT),
-            element(Tree.Kind.MEMBER_SELECT),
+        element(Tree.Kind.MEMBER_SELECT),
```
Comment (benzonico) [inline java-squid/src/test/java/org/sonar/java/cfg/CFGTest.java]: formatting

File: java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java [hunk 3]
```diff
@@ -65,19 +65,30 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
...
+
+  private ProgramState checkMemberSelect(CheckerContext context, MemberSelectExpressionTree syntaxNode, SymbolicValue currentVal) {
+    if ("class".equals(syntaxNode.identifier().name())) {
+      // expression ClassName.class won't raise NPE.
+      return context.getState();
+    }
 
-      if (context.isNull(currentVal)) {
+    if (context.isNull(currentVal)) {
         context.reportIssue(syntaxNode, this, "NullPointerException might be thrown as '" + SyntaxTreeNameFinder.getName(syntaxNode) + "' is nullable here");
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java]: MAJOR Make this line start at column 7. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AIndentationCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Pull request merged**